### PR TITLE
Add script to update SHA1 calculations of C and Python profiles.

### LINKED ISSUE #939: Add SHA1 update script
The `Generator` profiles use a SHA to identify whether the profile has been modified or not.  We should integrate a CMake custom target that will generate this to make updating the values in the code easier.

### PR DESCRIPTION
**Pull request by hsorby**
Fixes #939.

### CODE REVIEW

File: cmake/utils/CMakeLists.txt [hunk 1]
```diff
@@ -26,3 +26,17 @@ add_custom_command(OUTPUT mathmldtd.h
...
   DEPENDS mathmldtd.h
   COMMENT "Create mathmldtd files (in source directory, Eek!).")
+
+add_executable(calculatesha1 calculatesha1.cpp ${PROJECT_SOURCE_DIR}/src/profiletools.cpp)
+set_target_properties(calculatesha1 PROPERTIES EXCLUDE_FROM_DEFAULT_BUILD TRUE EXCLUDE_FROM_ALL TRUE)
+target_include_directories(calculatesha1 PRIVATE ${PROJECT_SOURCE_DIR}/src)
+target_link_libraries(calculatesha1 PRIVATE cellml)
+
+add_custom_command(OUTPUT profilesha1.h
+  COMMENT "Updating SHA1 values for C and Python profiles."
```
Comment (agarny) [inline cmake/utils/CMakeLists.txt]: "SHA-1", not "SHA1" (same below).

File: cmake/utils/calculatesha1.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+
```
Comment (agarny) [inline cmake/utils/calculatesha1.cpp:16]: Why that empty line? Also, we ought to have our copyright. Same actually for `cmake/utils/compress.cpp` (but not `cmake/utils/compress.in.h` which is fine).

File: cmake/utils/calculatesha1.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+#include <string>
+
+#include <libcellml/generatorprofile.h>
+
+#include "profiletools.h"
+#include "utilities.h"
+
+int main()
+{
+    auto c_profile = libcellml::GeneratorProfile::create();
```
Comment (agarny) [inline cmake/utils/calculatesha1.cpp]: C/C++ naming, not Python naming, i.e. `cProfile` not `c_profile`. Same for the other variables in this method.

File: src/profiletools.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * @brief Form the @c std::string representation of the given @p profile.
+ *
+ * Form the @c std::string representation of the given @p profile.
+ *
+ * @param profile The profile to create the @c std::string representation of.
+ *
+ * @return The @c std::string representation of the given @p profile.
+ */
+std::string formProfileRepresentation(const GeneratorProfilePtr &profile);
```
Comment (agarny) [inline src/profiletools.h]: I would rename this method `profileRepresentation()` or even `generatorProfileAsString()` (to be consiistent wiith `cellmlElementTypeAsString()`).

**Review by agarny — CHANGES_REQUESTED** [review]
A few minor changes.

This being said: I would REALLY like the SHA-1 values to be updated automatically, as we do with `versionconfig.h`. Right now, I have to confiigure CMake with `ENABLE_UTILITIES` set to `ON` and then I have to manually run the `update_profile_sha1_values` target. So, yeah, I can't see myself remembering to do that.

So, I will likely keep updating the SHA-1 values myself manually, as I have been doiong until now, except that the codebase is now giong to make things even less convenient since part of the code is in `src/generator.cpp` and the other in `profiletools.cpp` (incidentally, I would for consistency rename that file to `generatorprofiletools.cpp`).

**Bottom line:** this PR is supposed to make our life easier, but it won't make mine easier.

File: cmake/utils/calculatesha1.cpp (hunk 2 above)
Comment (hsorby) [reply cmake/utils/calculatesha1.cpp:16]: [be660ac](https://github.com/cellml/libcellml/pull/1054/commits/be660acec47a3ea93f2411ead69763664c4a43cd)

File: cmake/utils/calculatesha1.cpp (hunk 3 above)
Comment (hsorby) [reply cmake/utils/calculatesha1.cpp]: [58acedf](https://github.com/cellml/libcellml/pull/1054/commits/58acedf32f897f242fda6118c99ecae75d211287)

File: src/profiletools.h (hunk 4 above)
Comment (hsorby) [reply src/profiletools.h]: [3d7eeee](https://github.com/cellml/libcellml/pull/1054/commits/3d7eeeea7180f80ada3a4d3b3b818e97369d2ac1)

File: cmake/utils/CMakeLists.txt (hunk 1 above)
Comment (hsorby) [reply cmake/utils/CMakeLists.txt]: [f516667](https://github.com/cellml/libcellml/pull/1054/commits/f5166677b4cbff3235f675b7a584b8c62f1fafef)

File: cmake/utils/calculatesha1.cpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#include <string>
+
+#include <libcellml/generatorprofile.h>
+
+#include "generatorprofiletools.h"
+#include "utilities.h"
+
+int main()
+{
+    auto cProfile = libcellml::GeneratorProfile::create();
```
Comment (agarny) [inline cmake/utils/calculatesha1.cpp]: Might want to rename that variable to `cGeneratorProfile`. Also have `pyGeneratorProfile`, `cGeneratorProfileString`, and `pyGeneratorProfileString`.

File: src/generatorprofiletools.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * @brief Form the @c std::string representation of the given @p profile.
+ *
+ * Form the @c std::string representation of the given @p profile.
+ *
+ * @param profile The profile to create the @c std::string representation of.
+ *
+ * @return The @c std::string representation of the given @p profile.
+ */
+std::string generatorProfileAsString(const GeneratorProfilePtr &profile);
```
Comment (agarny) [inline src/generatorprofiletools.h]: `generatorProfile` rather than `profile`.

File: src/configure/generatorprofilesha1values.in.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+#include <string>
+
+namespace libcellml {
+
+/**
+ * This content of this file is generated, do not edit this file directly.
+ * See docs/dev_utilities.rst for further information.
+ */
+static const std::string C_PROFILE_SHA1 = "${C_PROFILE_SHA1_VALUE}";
+static const std::string PYTHON_PROFILE_SHA1 = "${PYTHON_PROFILE_SHA1_VALUE}";
```
Comment (agarny) [inline src/configure/generatorprofilesha1values.in.h]: Rename to `C_GENERATOR_PROFILE_SHA1` and `PYTHON_GENERATOR_PROFILE_SHA1`, or `C_GEN_PROF_SHA1` and `PY_GEN_PROF_SHA1` for a shorter version. Same in `src/generator.cpp`.

File: docs/dev_utilities.rst [hunk 8]
```diff
@@ -0,0 +1,31 @@
+=========
+Utilities
+=========
+
+This document covers the utilities that are availble to developers of libCellML.
+The utilities are not built or made available by default.
+The utilities can be enabled through the configuration of the library.
+In order to enable the utilities set the configuration variable LIBCELLML_ENABLE_UTILITES to true.
```
Comment (agarny) [inline docs/dev_utilities.rst]: "... utilities, set..."
Replace "LIBCELLML_ENABLE_UTILITES" with "`LIBCELLML_ENABLE_UTILITES`" and "true" with "`TRUE`".

File: docs/dev_utilities.rst [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+The available utilities are:
+
+#. Create MathML DTD files
+#. Update generator profile SHA-1 values
+
+Create MathML DTD Files
```
Comment (agarny) [inline docs/dev_utilities.rst]: "files", not "Files".

File: docs/dev_utilities.rst [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+=======================
+
+The create MathML DTD files utility generates a compressed MathML DTD of determining the validity of MathML strings.
```
Comment (agarny) [inline docs/dev_utilities.rst]: "... of determining..."? Might want to rephrase?

File: docs/dev_utilities.rst [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+It can be run using the *create_mathml_dtd_files* target.
```
Comment (agarny) [inline docs/dev_utilities.rst]: Prefer code block rather than italics for `create_mathml_dtd_files`.

File: docs/dev_utilities.rst [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+For a make based envirionment the utility may be run with the following command::
```
Comment (agarny) [inline docs/dev_utilities.rst]: "make-based"? (same below)
"... envirionment, the..."

File: docs/dev_utilities.rst [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+  make create_mathml_dtd_files
+
+Update generator profile SHA-1 values
+=====================================
+
+The update generator profile SHA-1 values utility calculates the SHA-1 values for the current C and Python profiles.
```
Comment (agarny) [inline docs/dev_utilities.rst]: "... Python generator profiles."

File: docs/dev_utilities.rst [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+The utility updates the header file *generatorprofilesha1values.h* in the source tree of the current build.
```
Comment (agarny) [inline docs/dev_utilities.rst]: Again, would use code blocks for `generatorprofilesha1values.h*`.

**Review by agarny — CHANGES_REQUESTED** [review]
Getting close, even though I still don't like the fact that I have to configure CMake in a specific way and then run a specific target to get the SHA-1 values updated. This should be done automatically for me. Simple as that. It's not like it takes sveral minutes to update those SHA-1 values.

File: docs/dev_utilities.rst (hunk 14 above)
Comment (hsorby) [reply docs/dev_utilities.rst]: [01d85ce](https://github.com/cellml/libcellml/pull/1054/commits/01d85ce541a16e5170eb41abe1212977574e2b51)

File: src/configure/generatorprofilesha1values.in.h (hunk 7 above)
Comment (hsorby) [reply src/configure/generatorprofilesha1values.in.h]: [e6a1541](https://github.com/cellml/libcellml/pull/1054/commits/e6a1541914edc61403663ced7d23188e6dbb5a16)

File: src/generatorprofiletools.h (hunk 6 above)
Comment (hsorby) [reply src/generatorprofiletools.h]: [1ccba6a](https://github.com/cellml/libcellml/pull/1054/commits/1ccba6ade014862ec984452997d60759250019af)

File: cmake/utils/calculatesha1.cpp (hunk 5 above)
Comment (hsorby) [reply cmake/utils/calculatesha1.cpp]: [86a9d00](https://github.com/cellml/libcellml/pull/1054/commits/86a9d006aa0d4cd9108a57e6ef42ce7652abd9f7)

File: docs/dev_utilities.rst [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+The available utilities are:
+
+#. Create MathML DTD files
+#. Update generator profile SHA-1 values
+
+Create MathML DTD files
+=======================
+
+The create MathML DTD files utility generates a compressed MathML DTD.
+The MathML DTD is used to determine the validity of MathML strings in a CellML model.
```
Comment (agarny) [inline docs/dev_utilities.rst]: "... CellML document."

Comment (hsorby) [reply docs/dev_utilities.rst]: [f6d85c5](https://github.com/cellml/libcellml/pull/1054/commits/f6d85c5f8cd1162364a6d4fc7ed143e4ee13a704)

**Review by agarny — APPROVED** [review]

**Review by nickerso — APPROVED** [review]
assuming tests pass...this looks good to me.
